File: packages/provide-uterm-csharp/src/Provide.Uterm/ControlChannel/Codec.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

using System.Globalization;
using System.Text;
using System.Text.Json;

namespace Provide.Uterm.ControlChannel;

/// <summary>
/// Inline DLE/STX control framing used to mix terminal data and JSON control
/// messages in one WebSocket stream. Port of provide.uterm.control_channel /
/// packages/provide-uterm-go/controlchannel.
///
/// Internally the decoder buffers UTF-8 bytes (matching Go string semantics).
/// Public APIs accept/return .NET strings via UTF-8 round-trip.
/// </summary>
public static class ControlChannelCodec
{
    public const string Dle = "\u0010";
    public const string Stx = "\u0002";
    public const byte DleByte = 0x10;
    public const byte StxByte = 0x02;

    internal const int HeaderBytes = 11; // DLE STX + 8 hex digits + ':'
    internal const int MaxControlPayloadBytes = 1_048_576;
    internal const int DefaultMaxBufferBytes = 10_485_760;
    internal const int DefaultMaxFrameDepth = 32;

    /// <summary>Encode terminal data by escaping every DLE byte.</summary>
    public static string EncodeTerminalData(string data) =>
        data.Replace(Dle, Dle + Dle, StringComparison.Ordinal);

    /// <summary>
    /// Encode a control payload. JSON is compact (no spaces) and does not
    /// escape HTML metacharacters, matching the Python encoder.
    /// </summary>
    public static string EncodeControlFrame(IReadOnlyDictionary<string, object?> payload)
    {
        var serialized = MarshalCompact(payload);
        var utf8Len = Encoding.UTF8.GetByteCount(serialized);
        return $"{Dle}{Stx}{utf8Len:x8}:{serialized}";
    }

    public static string EncodeControlFrame(Dictionary<string, object?> payload) =>
        EncodeControlFrame((IReadOnlyDictionary<string, object?>)payload);

    internal static string MarshalCompact(IReadOnlyDictionary<string, object?> payload)
    {
        using var stream = new MemoryStream();
        using (var writer = new Utf8JsonWriter(stream, new JsonWriterOptions
        {
            Indented = false,
            Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
        }))
        {
            WriteValue(writer, payload);
        }

        return Encoding.UTF8.GetString(stream.ToArray());
    }

    private static void WriteValue(Utf8JsonWriter writer, object? value)
    {
        switch (value)
        {
            case null:
                writer.WriteNullValue();
                break;
            case bool b:
                writer.WriteBooleanValue(b);
                break;
            case string s:
                writer.WriteStringValue(s);
                break;
            case byte or sbyte or short or ushort or int or uint or long or ulong:
                writer.WriteNumberValue(Convert.ToInt64(value, CultureInfo.InvariantCulture));
                break;
            case float f:
                writer.WriteNumberValue(f);
                break;
            case double d:
                writer.WriteNumberValue(d);
                break;
            case decimal m:
                writer.WriteNumberValue(m);
                break;
            case JsonElement je:
                je.WriteTo(writer);
                break;
            case IReadOnlyDictionary<string, object?> dict:
                writer.WriteStartObject();
                foreach (var (k, v) in dict)
                {
                    writer.WritePropertyName(k);
                    WriteValue(writer, v);
                }

                writer.WriteEndObject();
                break;
            case IDictionary<string, object?> dict:
                writer.WriteStartObject();
                foreach (var (k, v) in dict)
                {
                    writer.WritePropertyName(k);
                    WriteValue(writer, v);
                }

                writer.WriteEndObject();
                break;
            case System.Collections.IEnumerable list when value is not string:
                writer.WriteStartArray();
                foreach (var item in list)
                {
                    WriteValue(writer, item);
                }

                writer.WriteEndArray();
                break;
            default:
                writer.WriteStringValue(Convert.ToString(value, CultureInfo.InvariantCulture));
                break;
        }
    }

    public static Dictionary<string, object?> JsonElementToDictionary(JsonElement el)
    {
        var dict = new Dictionary<string, object?>();
        foreach (var prop in el.EnumerateObject())
        {
            dict[prop.Name] = JsonElementToObject(prop.Value);
        }

        return dict;
    }

    /// <summary>
    /// Convert JSON to CLR objects, preserving integer vs float distinction
    /// (integers stay long; fractional values stay double) for HMAC canonicalization.
    /// </summary>
    internal static object? JsonElementToObject(JsonElement el) =>
        el.ValueKind switch
        {
            JsonValueKind.Object => JsonElementToDictionary(el),
            JsonValueKind.Array => el.EnumerateArray().Select(JsonElementToObject).Cast<object?>().ToList(),
            JsonValueKind.String => el.GetString(),
            JsonValueKind.Number => NumberFromJson(el),
            JsonValueKind.True => true,
            JsonValueKind.False => false,
            JsonValueKind.Null => null,
            _ => el.GetRawText(),
        };

    private static object NumberFromJson(JsonElement el)
    {
        var raw = el.GetRawText();
        if (raw.IndexOfAny(['.', 'e', 'E']) < 0 && el.TryGetInt64(out var l))
        {
            return l;
        }

        return el.GetDouble();
    }

    /// <summary>
    /// Structural check: magic, length header syntax, and full payload present.
    /// </summary>
    public static bool IsControlFrame(string message)
    {
        var bytes = Encoding.UTF8.GetBytes(message);
        return IsControlFrame(bytes);
    }

    /// <summary>Structural control-frame check over the original wire bytes.</summary>
    public static bool IsControlFrame(ReadOnlySpan<byte> bytes)
    {
        if (bytes.Length < HeaderBytes)
        {
            return false;
        }

        if (bytes[0] != DleByte || bytes[1] != StxByte)
        {
            return false;
        }

        if (bytes[10] != (byte)':')
        {
            return false;
        }

        var lengthHex = Encoding.ASCII.GetString(bytes.Slice(2, 8));
        if (!TryParseHex32(lengthHex, out var payloadBytes))
        {
            return false;
        }

        if (string.Create(CultureInfo.InvariantCulture, $"{payloadBytes:x8}") != lengthHex)
        {
            return false;
        }

        if (payloadBytes > MaxControlPayloadBytes)
        {
            return false;
        }

        try
        {
            // Narrowed only after the size guard above, for the reason in
            // TryParseHex32: the header is unsigned and this argument is not.
            var payloadEnd = Utf8PayloadEnd(bytes, HeaderBytes, (int)payloadBytes);
            return payloadEnd == bytes.Length;
        }
        catch (ProtocolException)
        {
            return false;
        }
    }

    /// <summary>
    /// Parse the frame's eight-digit hex length header.
    /// </summary>
    /// <remarks>
    /// The accumulator is <see cref="long"/>, not <see cref="int"/>, and that is
    /// the whole point: the header is an <em>unsigned</em> 32-bit wire value, so
    /// <c>80000000</c> through <c>ffffffff</c> overflow a signed 32-bit
    /// accumulator and wrap negative. A negative length then slips past every
    /// <c>&gt; MaxControlPayloadBytes</c> guard — which only ever looks upward —
    /// and reaches an index or a slice, where it throws a raw
    /// <see cref="IndexOutOfRangeException"/> instead of the protocol error a
    /// caller is catching. Thirteen bytes from a peer were enough. The reference
    /// parses the same header with arbitrary-precision arithmetic and reports
    /// "control payload too large"; pinned as <c>CCF-REG-0005</c> in the
    /// cross-language fuzz corpus, which is what found this.
    ///
    /// Callers must therefore keep the value wide until after the size guard,
    /// and only then narrow it.
    /// </remarks>
    internal static bool TryParseHex32(ReadOnlySpan<char> s, out long value)
    {
        value = 0;
        if (s.Length != 8)
        {
            return false;
        }

        for (var i = 0; i < s.Length; i++)
        {
            var c = s[i];
            int d;
            if (c is >= '0' and <= '9')
            {
                d = c - '0';
            }
            else if (c is >= 'a' and <= 'f')
            {
                d = c - 'a' + 10;
            }
            else if (c is >= 'A' and <= 'F')
            {
                d = c - 'A' + 10;
            }
            else
            {
                return false;
            }

            value = (value << 4) | d;
        }

        return true;
    }

    /// <summary>
    /// Byte index ending a payload of <paramref name="payloadBytes"/> UTF-8
    /// bytes starting at <paramref name="start"/>, or -1 when incomplete.
    /// Throws when the declared length splits a multi-byte rune.
    /// </summary>
    internal static int Utf8PayloadEnd(ReadOnlySpan<byte> buf, int start, int payloadBytes)
    {
        var end = start + payloadBytes;
        if (end > buf.Length)
        {
            return -1;
        }

        if (end < buf.Length && IsUtf8Continuation(buf[end]))
        {
            throw new ProtocolException("invalid control payload length");
        }

        return end;
    }

    private static bool IsUtf8Continuation(byte b) => (b & 0xC0) == 0x80;
}

/// <summary>Raised when an inline control frame is malformed.</summary>
public sealed class ProtocolException : Exception
{
    public ProtocolException(string message)
        : base(message)
    {
    }
}

/// <summary>Decoded element of the inline stream.</summary>
public abstract class Chunk
{
    public abstract string Kind { get; }
}

/// <summary>Decoded terminal data (UTF-8 decoded string, latin-1 safe for BBS).</summary>
public sealed class DataChunk : Chunk
{
    public DataChunk(string data) => Data = data;

    public string Data { get; }

    public override string Kind => "data";
}

/// <summary>Decoded control payload.</summary>
public sealed class ControlChunk : Chunk
{
    public ControlChunk(Dictionary<string, object?> control) => Control = control;

    public Dictionary<string, object?> Control { get; }

    /// <summary>Alias for <see cref="Control"/> (some call sites use Payload).</summary>
    public Dictionary<string, object?> Payload => Control;

    public override string Kind => "control";
}

/// <summary>Decoder options. Zero/defaults select protocol defaults.</summary>
public sealed class DecoderOptions
{
    public int MaxControlPayloadBytes { get; set; }
    public int MaxBufferBytes { get; set; }
    public int MaxFrameDepth { get; set; }
    public Action<string>? OnError { get; set; }
}

/// <summary>Incrementally decodes the inline DLE/STX control-frame stream.</summary>
public class ControlFrameDecoder
{
    private readonly int _maxControlPayloadBytes;
    private readonly int _maxBufferBytes;
    private readonly int _maxFrameDepth;
    private readonly Action<string>? _onError;
    private byte[] _buffered = Array.Empty<byte>();
    private int _bufferedLen;

    public ControlFrameDecoder(DecoderOptions? options = null)
    {
        options ??= new DecoderOptions();
        _maxControlPayloadBytes = options.MaxControlPayloadBytes < 1
            ? ControlChannelCodec.MaxControlPayloadBytes
            : options.MaxControlPayloadBytes;
        _maxBufferBytes = options.MaxBufferBytes < 1
            ? ControlChannelCodec.DefaultMaxBufferBytes
            : options.MaxBufferBytes;
        _maxFrameDepth = options.MaxFrameDepth < 1
            ? ControlChannelCodec.DefaultMaxFrameDepth
            : options.MaxFrameDepth;
        _onError = options.OnError;
    }

    private Exception ReportError(string message)
    {
        _onError?.Invoke("control_frame_protocol_error");
        return new ProtocolException(message);
    }

    private void Reset()
    {
        _buffered = Array.Empty<byte>();
        _bufferedLen = 0;
    }

    private void Append(ReadOnlySpan<byte> chunk)
    {
        if (_bufferedLen + chunk.Length > _buffered.Length)
        {
            var next = Math.Max(_buffered.Length == 0 ? 256 : _buffered.Length * 2, _bufferedLen + chunk.Length);
            Array.Resize(ref _buffered, next);
        }

        chunk.CopyTo(_buffered.AsSpan(_bufferedLen));
        _bufferedLen += chunk.Length;
    }

    /// <summary>Decode all complete events from <paramref name="chunk"/>.</summary>
    public IReadOnlyList<Chunk> Feed(string chunk)
    {
        var bytes = Encoding.UTF8.GetBytes(chunk);
        return FeedBytes(bytes);
    }

    /// <summary>
    /// Decode directly from wire bytes. Binary terminal data can request a
    /// one-byte-to-one-character mapping while control JSON remains UTF-8.
    /// </summary>
    public IReadOnlyList<Chunk> FeedBytes(ReadOnlySpan<byte> chunk, bool preserveRawData = false)
    {
        if (_bufferedLen + chunk.Length > _maxBufferBytes)
        {
            var total = _bufferedLen + chunk.Length;
            Reset();
            throw ReportError($"control frame buffer overflow: {total} > {_maxBufferBytes}");
        }

        Append(chunk);
        try
        {
            return Drain(final: false, preserveRawData: preserveRawData);
        }
        catch
        {
            Reset();
            throw;
        }
    }

    /// <summary>Decode remaining buffered data; reject truncated frames.</summary>
    public IReadOnlyList<Chunk> Finish()
    {
        try
        {
            return Drain(final: true, preserveRawData: false);
        }
        catch
        {
            Reset();
            throw;
        }
    }

    private Dictionary<string, object?> ParseFramePayload(ReadOnlySpan<byte> payloadRaw)
    {
        JsonDocument doc;
        try
        {
            doc = JsonDocument.Parse(payloadRaw.ToArray());
        }
        catch (JsonException)
        {
            throw ReportError("invalid control json");
        }

        using (doc)
        {
            if (doc.RootElement.ValueKind != JsonValueKind.Object)
            {
                throw ReportError("control payload must be an object");
            }

            var obj = ControlChannelCodec.JsonElementToDictionary(doc.RootElement);
            CheckJsonDepth(obj, _maxFrameDepth);
            return obj;
        }
    }

    private void CheckJsonDepth(object? value, int maxDepth)
    {
        var stack = new Stack<(object? Node, int Depth)>();
        stack.Push((value, 1));
        while (stack.Count > 0)
        {
            var (node, depth) = stack.Pop();
            if (depth > maxDepth)
            {
                throw ReportError($"control payload nests deeper than {maxDepth}");
            }

            switch (node)
            {
                case Dictionary<string, object?> map:
                    foreach (var child in map.Values)
                    {
                        if (child is Dictionary<string, object?> or System.Collections.IList)
                        {
                            stack.Push((child, depth + 1));
                        }
                    }

                    break;
                case System.Collections.IList list:
                    foreach (var child in list)
                    {
                        if (child is Dictionary<string, object?> or System.Collections.IList)
                        {
                            stack.Push((child, depth + 1));
                        }
                    }

                    break;
            }
        }
    }

    private (ControlChunk? Chunk, int FrameEnd, bool Done) TryParseFrame(ReadOnlySpan<byte> buf, int idx, bool final)
    {
        if (buf.Length - idx < ControlChannelCodec.HeaderBytes)
        {
            if (final)
            {
                throw ReportError("truncated control frame");
            }

            return (null, 0, false);
        }

        var lengthHex = Encoding.ASCII.GetString(buf.Slice(idx + 2, 8));
        var separator = buf[idx + 10];
        if (separator != (byte)':' || !ControlChannelCodec.TryParseHex32(lengthHex, out var payloadBytes))
        {
            throw ReportError("invalid control header");
        }

        if (payloadBytes > ControlChannelCodec.MaxControlPayloadBytes || payloadBytes > _maxControlPayloadBytes)
        {
            throw ReportError("control payload too large");
        }

        var payloadStart = idx + ControlChannelCodec.HeaderBytes;
        // Narrowed only here, after both size guards above have run. The header
        // is an unsigned 32-bit value; narrowing before the guard is what let a
        // negative length reach this index. See TryParseHex32.
        var payloadLength = (int)payloadBytes;
        int end;
        try
        {
            end = ControlChannelCodec.Utf8PayloadEnd(buf, payloadStart, payloadLength);
        }
        catch (ProtocolException ex)
        {
            throw ReportError(ex.Message);
        }

        if (end == -1)
        {
            if (final)
            {
                throw ReportError("truncated control frame");
            }

            return (null, 0, false);
        }

        var payload = ParseFramePayload(buf.Slice(payloadStart, payloadLength));
        return (new ControlChunk(payload), end, true);
    }

    private List<Chunk> Drain(bool final, bool preserveRawData)
    {
        var events = new List<Chunk>();
        // Copy to array so local functions can capture without ref-struct issues.
        var buf = new byte[_bufferedLen];
        Buffer.BlockCopy(_buffered, 0, buf, 0, _bufferedLen);
        var bufLen = buf.Length;
        var idx = 0;
        var dataParts = new List<byte>();
        var dataStart = 0;

        void EmitData(int upTo)
        {
            if (dataStart < upTo)
            {
                for (var i = dataStart; i < upTo; i++)
                {
                    dataParts.Add(buf[i]);
                }
            }

            if (dataParts.Count > 0)
            {
                var data = dataParts.ToArray();
                events.Add(new DataChunk(preserveRawData
                    ? Encoding.Latin1.GetString(data)
                    : Encoding.UTF8.GetString(data)));
                dataParts.Clear();
            }
        }

        while (idx < bufLen)
        {
            if (buf[idx] != ControlChannelCodec.DleByte)
            {
                idx++;
                continue;
            }

            if (idx + 1 >= bufLen)
            {
                if (final)
                {
                    throw ReportError("truncated control frame");
                }

                break;
            }

            var next = buf[idx + 1];
            if (next == ControlChannelCodec.DleByte)
            {
                if (dataStart < idx)
                {
                    for (var i = dataStart; i < idx; i++)
                    {
                        dataParts.Add(buf[i]);
                    }
                }

                dataParts.Add(ControlChannelCodec.DleByte);
                idx += 2;
                dataStart = idx;
                continue;
            }

            if (next != ControlChannelCodec.StxByte)
            {
                throw ReportError("invalid control prefix");
            }

            EmitData(idx);
            dataStart = idx;

            var (chunk, frameEnd, done) = TryParseFrame(buf, idx, final);
            if (!done)
            {
                break;
            }

            idx = frameEnd;
            dataStart = idx;
            if (chunk is not null)
            {
                events.Add(chunk);
            }
        }

        if (idx > 0)
        {
            var remaining = _bufferedLen - idx;
            if (remaining > 0)
            {
                Buffer.BlockCopy(_buffered, idx, _buffered, 0, remaining);
            }

            _bufferedLen = remaining;
        }

        EmitData(idx);
        return events;
    }
}

/// <summary>Go/Python-aligned name for <see cref="ControlFrameDecoder"/>.</summary>
public sealed class Decoder : ControlFrameDecoder
{
    public Decoder(DecoderOptions? options = null) : base(options) { }
}

File: packages/provide-uterm-go/controlchannel/codec.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

// Package controlchannel implements the inline DLE/STX control framing used to
// mix terminal data and JSON control messages in one WebSocket stream.
//
// This is a direct port of provide.uterm.control_channel (Python). Control
// frame headers store the UTF-8 byte length of the JSON payload. Buffers are
// Go strings (UTF-8), so the declared payload byte length maps directly onto
// string byte offsets; a declared length that splits a multi-byte rune is a
// protocol error, exactly as the Python decoder treats a length that splits a
// Unicode code point.
package controlchannel

import (
	"bytes"
	"encoding/json"
	"fmt"
	"strings"
	"unicode/utf8"
)

const (
	// DLE and STX are the frame magic bytes.
	DLE = "\x10"
	STX = "\x02"

	headerBytes            = 11 // DLE STX + 8 hex digits + ':'
	maxControlPayloadBytes = 1_048_576
	defaultMaxBufferBytes  = 10_485_760
	defaultMaxFrameDepth   = 32
	dleByte                = 0x10
	stxByte                = 0x02
)

// ProtocolError is raised when an inline control frame is malformed.
type ProtocolError struct {
	msg string
}

func (e *ProtocolError) Error() string { return e.msg }

func protocolErrorf(format string, args ...any) *ProtocolError {
	return &ProtocolError{msg: fmt.Sprintf(format, args...)}
}

// Chunk is a decoded element of the inline stream: either terminal data or a
// control payload.
type Chunk interface {
	// Kind returns "data" or "control".
	Kind() string
}

// DataChunk is decoded terminal data from the inline stream.
type DataChunk struct {
	Data string
}

// Kind implements Chunk.
func (DataChunk) Kind() string { return "data" }

// ControlChunk is a decoded control payload from the inline stream.
type ControlChunk struct {
	Control map[string]any
}

// Kind implements Chunk.
func (ControlChunk) Kind() string { return "control" }

// EncodeTerminalData encodes terminal data for the inline stream by escaping
// every DLE byte.
func EncodeTerminalData(data string) string {
	return strings.ReplaceAll(data, DLE, DLE+DLE)
}

// EncodeControlFrame encodes a control payload for the inline stream.
//
// The JSON serialization is compact (no spaces) and does not escape non-ASCII
// characters or HTML metacharacters, matching the Python encoder.
func EncodeControlFrame(payload map[string]any) (string, error) {
	serialized, err := marshalCompact(payload)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%s%s%08x:%s", DLE, STX, len(serialized), serialized), nil
}

func marshalCompact(payload map[string]any) (string, error) {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(payload); err != nil {
		return "", err
	}
	// json.Encoder appends a trailing newline; the wire format must not carry it.
	return strings.TrimSuffix(buf.String(), "\n"), nil
}

func parseHex32(s string) (int, bool) {
	v := 0
	for i := 0; i < len(s); i++ {
		c := s[i]
		var d int
		switch {
		case c >= '0' && c <= '9':
			d = int(c - '0')
		case c >= 'a' && c <= 'f':
			d = int(c-'a') + 10
		case c >= 'A' && c <= 'F':
			d = int(c-'A') + 10
		default:
			return 0, false
		}
		v = v<<4 | d
	}
	return v, true
}

// utf8PayloadEnd returns the byte index ending a payload of payloadBytes UTF-8
// bytes starting at start, or -1 when buf does not yet contain that many
// bytes. It returns a ProtocolError when the declared byte length splits a
// multi-byte rune (which appending more data cannot fix).
func utf8PayloadEnd(buf string, start, payloadBytes int) (int, error) {
	end := start + payloadBytes
	if end > len(buf) {
		return -1, nil
	}
	// Go strings from the WebSocket layer are valid UTF-8, so end == len(buf)
	// is always a rune boundary; otherwise the byte at end must not be a
	// UTF-8 continuation byte.
	if end < len(buf) && !utf8.RuneStart(buf[end]) {
		return -1, protocolErrorf("invalid control payload length")
	}
	return end, nil
}

// IsControlFrame reports whether message is a full control-framed payload.
//
// The check is structural only: it validates the magic bytes, length header
// syntax, and that the declared UTF-8 payload bytes are fully present.
func IsControlFrame(message string) bool {
	if len(message) < headerBytes {
		return false
	}
	if !strings.HasPrefix(message, DLE+STX) {
		return false
	}
	if message[10] != ':' {
		return false
	}
	lengthHex := message[2:10]
	payloadBytes, ok := parseHex32(lengthHex)
	if !ok {
		return false
	}
	if fmt.Sprintf("%08x", payloadBytes) != lengthHex {
		return false
	}
	if payloadBytes > maxControlPayloadBytes {
		return false
	}
	payloadEnd, err := utf8PayloadEnd(message, headerBytes, payloadBytes)
	if err != nil {
		return false
	}
	return payloadEnd != -1 && payloadEnd == len(message)
}

// checkJSONDepth returns a ProtocolError if value nests deeper than maxDepth.
// Maps and slices each add one level; primitive leaves count as depth 0. The
// walk is iterative so a pathological payload cannot exhaust the stack.
func checkJSONDepth(value any, maxDepth int) error {
	type item struct {
		node  any
		depth int
	}
	stack := []item{{value, 1}}
	for len(stack) > 0 {
		it := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if it.depth > maxDepth {
			return protocolErrorf("control payload nests deeper than %d", maxDepth)
		}
		switch node := it.node.(type) {
		case map[string]any:
			for _, child := range node {
				switch child.(type) {
				case map[string]any, []any:
					stack = append(stack, item{child, it.depth + 1})
				}
			}
		case []any:
			for _, child := range node {
				switch child.(type) {
				case map[string]any, []any:
					stack = append(stack, item{child, it.depth + 1})
				}
			}
		}
	}
	return nil
}

// DecoderOptions configure a Decoder. Zero values select the defaults.
type DecoderOptions struct {
	// MaxControlPayloadBytes bounds a single control frame payload. Values
	// above the protocol maximum (1 MiB) are still capped at 1 MiB.
	MaxControlPayloadBytes int
	// MaxBufferBytes bounds the total undecoded buffer.
	MaxBufferBytes int
	// MaxFrameDepth bounds JSON nesting inside a control frame.
	MaxFrameDepth int
	// OnError, when set, is invoked with "control_frame_protocol_error"
	// whenever the decoder raises a protocol error.
	OnError func(string)
}

// Decoder incrementally decodes the inline DLE/STX control-frame stream.
type Decoder struct {
	maxControlPayloadBytes int
	maxBufferBytes         int
	maxFrameDepth          int
	onError                func(string)
	buffered               string
}

// NewDecoder creates a Decoder with the given options.
func NewDecoder(opts DecoderOptions) *Decoder {
	d := &Decoder{
		maxControlPayloadBytes: opts.MaxControlPayloadBytes,
		maxBufferBytes:         opts.MaxBufferBytes,
		maxFrameDepth:          opts.MaxFrameDepth,
		onError:                opts.OnError,
	}
	if d.maxControlPayloadBytes < 1 {
		d.maxControlPayloadBytes = maxControlPayloadBytes
	}
	if d.maxBufferBytes < 1 {
		d.maxBufferBytes = defaultMaxBufferBytes
	}
	if d.maxFrameDepth < 1 {
		d.maxFrameDepth = defaultMaxFrameDepth
	}
	return d
}

func (d *Decoder) reportError(format string, args ...any) error {
	if d.onError != nil {
		d.onError("control_frame_protocol_error")
	}
	return protocolErrorf(format, args...)
}

func (d *Decoder) reset() {
	d.buffered = ""
}

// Feed decodes all complete events from chunk and buffers the rest.
func (d *Decoder) Feed(chunk string) ([]Chunk, error) {
	if len(d.buffered)+len(chunk) > d.maxBufferBytes {
		total := len(d.buffered) + len(chunk)
		d.reset()
		return nil, d.reportError("control frame buffer overflow: %d > %d", total, d.maxBufferBytes)
	}
	d.buffered += chunk
	events, err := d.drain(false)
	if err != nil {
		d.reset()
		return nil, err
	}
	return events, nil
}

// Finish decodes any remaining buffered data and rejects truncated control
// frames.
func (d *Decoder) Finish() ([]Chunk, error) {
	events, err := d.drain(true)
	if err != nil {
		d.reset()
		return nil, err
	}
	// A final drain either consumes the whole buffer or errors: every
	// incomplete construct (lone DLE, short header, short payload) raises
	// "truncated control frame" inside drain, so no leftover check is needed.
	return events, nil
}

func (d *Decoder) parseFramePayload(payloadRaw string) (map[string]any, error) {
	dec := json.NewDecoder(strings.NewReader(payloadRaw))
	// Preserve each JSON number's source text as json.Number rather than
	// collapsing to float64. Python's json.loads keeps the int/float
	// distinction ("1" -> int, "1.0" -> float), which is load-bearing for
	// identity-frame HMAC signatures: a verifier must canonicalize the
	// received claims exactly as the producer did. float64 loses that
	// distinction (both decode to 1.0), so an integer claim would re-sign as
	// "1.0" and fail verification against Python's "1".
	dec.UseNumber()
	var payload any
	if err := dec.Decode(&payload); err != nil {
		return nil, d.reportError("invalid control json")
	}
	// Reject trailing garbage after the JSON document, matching Python's
	// strict json.loads behavior.
	if dec.More() {
		return nil, d.reportError("invalid control json")
	}
	obj, ok := payload.(map[string]any)
	if !ok {
		return nil, d.reportError("control payload must be an object")
	}
	if err := checkJSONDepth(obj, d.maxFrameDepth); err != nil {
		return nil, d.reportError("%s", err.Error())
	}
	return obj, nil
}

// tryParseFrame parses a control frame at buf[idx]. It returns the chunk and
// the frame end offset, or done=false when the frame is not yet complete
// (only valid when final is false).
func (d *Decoder) tryParseFrame(buf string, idx int, final bool) (chunk ControlChunk, frameEnd int, done bool, err error) {
	if len(buf)-idx < headerBytes {
		if final {
			return ControlChunk{}, 0, false, d.reportError("truncated control frame")
		}
		return ControlChunk{}, 0, false, nil
	}
	lengthHex := buf[idx+2 : idx+10]
	separator := buf[idx+10]
	payloadBytes, hexOK := parseHex32(lengthHex)
	// The canonical comparison, not just parseHex32's digit test: parseHex32
	// accepts A-F, so without this the decoder reads "0000001F" as a frame
	// while IsControlFrame -- which has always compared against %08x -- reads
	// the same bytes as terminal data. IsControlFrame is what decides whether
	// a message is framed at all, so the two must agree. Pinned as
	// CCF-REG-0006 in the shared fuzz corpus.
	if separator != ':' || !hexOK || fmt.Sprintf("%08x", payloadBytes) != lengthHex {
		return ControlChunk{}, 0, false, d.reportError("invalid control header")
	}
	// Bounding check: oversized frames are rejected before allocation.
	if payloadBytes > maxControlPayloadBytes || payloadBytes > d.maxControlPayloadBytes {
		return ControlChunk{}, 0, false, d.reportError("control payload too large")
	}
	payloadStart := idx + headerBytes
	end, perr := utf8PayloadEnd(buf, payloadStart, payloadBytes)
	if perr != nil {
		return ControlChunk{}, 0, false, d.reportError("%s", perr.Error())
	}
	if end == -1 {
		if final {
			return ControlChunk{}, 0, false, d.reportError("truncated control frame")
		}
		return ControlChunk{}, 0, false, nil
	}
	payload, err := d.parseFramePayload(buf[payloadStart:end])
	if err != nil {
		return ControlChunk{}, 0, false, err
	}
	return ControlChunk{Control: payload}, end, true, nil
}

func (d *Decoder) drain(final bool) ([]Chunk, error) {
	var events []Chunk
	buf := d.buffered
	bufLen := len(buf)
	idx := 0
	var dataParts strings.Builder
	dataStart := 0

	emitData := func(upTo int) {
		if dataStart < upTo {
			dataParts.WriteString(buf[dataStart:upTo])
		}
		if dataParts.Len() > 0 {
			events = append(events, DataChunk{Data: dataParts.String()})
			dataParts.Reset()
		}
	}

	for idx < bufLen {
		if buf[idx] != dleByte {
			idx++
			continue
		}
		if idx+1 >= bufLen {
			if final {
				return nil, d.reportError("truncated control frame")
			}
			break
		}
		next := buf[idx+1]
		if next == dleByte {
			// Escaped DLE: save the data slice before it, add a literal DLE.
			if dataStart < idx {
				dataParts.WriteString(buf[dataStart:idx])
			}
			dataParts.WriteString(DLE)
			idx += 2
			dataStart = idx
			continue
		}
		if next != stxByte {
			return nil, d.reportError("invalid control prefix")
		}

		emitData(idx)
		dataStart = idx

		chunk, frameEnd, done, err := d.tryParseFrame(buf, idx, final)
		if err != nil {
			return nil, err
		}
		if !done {
			break
		}
		idx = frameEnd
		dataStart = idx
		events = append(events, chunk)
	}

	// Flush the unconsumed buffer tail and any trailing plain data.
	if idx > 0 {
		d.buffered = buf[idx:]
	}
	emitData(idx)
	return events, nil
}
